src/main/scala/vsys/state/opcdiffs/TDBROpcDiff.scala

```diff
@@ -0,0 +1,162 @@
+    } else {
+      val contractTokens = context.state.contractTokens(context.contractId.bytes)
+      val tokenNumber = Ints.fromByteArray(tokenIndex.data)
+      val tokenID: ByteStr = tokenIdFromBytes(context.contractId.bytes.arr, tokenIndex.data).right.get
+      val tokenDescKey = ByteStr(Bytes.concat(tokenID.arr, Array(3.toByte)))
+      if (tokenNumber >= contractTokens || tokenNumber < 0) {
+        Left(ContractInvalidTokenIndex)
+      } else {
+        context.state.tokenInfo(tokenDescKey) match {
+          case Some(v) => Right(dataStack.patch(pointer, Seq(v), 1))
+          case _ => Left(ContractInvalidTokenInfo)
+        }
+      }
+    }
+  }
+
+  def descWithoutTokenIndex(context: ExecutionContext)(dataStack: Seq[DataEntry], pointer: Byte): Either[ValidationError, Seq[DataEntry]] = {
+
+    val tokenIndex = DataEntry(Ints.toByteArray(0), DataType.Int32)
+    desc(context)(tokenIndex, dataStack, pointer)
+  }
+
+  object TDBRType extends Enumeration {
+    val MaxTDBR = Value(1)
+    val TotalTDBR = Value(2)
+    val UnityTDBR = Value(3)
+    val DescTDBR = Value(4)
+  }
+
+  def parseBytes(context: ExecutionContext)
+                (bytes: Array[Byte], data: Seq[DataEntry]): Either[ValidationError, Seq[DataEntry]] = bytes.head match {
+    case opcType: Byte if opcType == TDBRType.MaxTDBR.id && checkInput(bytes.slice(0, bytes.length - 1),1, data.length) =>
+      maxWithoutTokenIndex(context)(data, bytes(1))
+    case opcType: Byte if opcType == TDBRType.MaxTDBR.id && checkInput(bytes.slice(0, bytes.length - 1),2, data.length) =>
+      max(context)(data(bytes(1)), data, bytes(2))
+    case opcType: Byte if opcType == TDBRType.TotalTDBR.id && checkInput(bytes.slice(0, bytes.length - 1),1, data.length) =>
+      totalWithoutTokenIndex(context)(data, bytes(1))
+    case opcType: Byte if opcType == TDBRType.TotalTDBR.id && checkInput(bytes.slice(0, bytes.length - 1),2, data.length) =>
+      total(context)(data(bytes(1)), data, bytes(2))
+    case opcType: Byte if opcType == TDBRType.UnityTDBR.id && checkInput(bytes.slice(0, bytes.length - 1),1, data.length) =>
+      unityWithoutTokenIndex(context)(data, bytes(1))
+    case opcType: Byte if opcType == TDBRType.UnityTDBR.id && checkInput(bytes.slice(0, bytes.length - 1),2, data.length) =>
+      unity(context)(data(bytes(1)), data, bytes(2))
+    case opcType: Byte if opcType == TDBRType.DescTDBR.id && checkInput(bytes.slice(0, bytes.length - 1),1, data.length) =>
+      descWithoutTokenIndex(context)(data, bytes(1))
+    case opcType: Byte if opcType == TDBRType.DescTDBR.id && checkInput(bytes.slice(0, bytes.length - 1),2, data.length) =>
+      desc(context)(data(bytes(1)), data, bytes(2))
+    case _ => Left(ContractInvalidOPCData)
+  }
+
+  private def checkInput(bytes: Array[Byte], bLength: Int, dataLength: Int): Boolean = {
+    bytes.length == bLength && (bytes.length == 1 || (bytes.tail.max < dataLength && bytes.tail.min >= 0))
```

Comment:
what is in bytes.tail? why should bytes.tail.max < dataLength && bytes.tail.min >= 0?

Reply:
bytes.tail represent the data entry index in data (Seq[DataEntry]), so it cannot accept an index out of range

Reply:
i see. thanks